MessageBusBackend improvements

ZipFile opened the pull request:
* Split buffering and aggregation features onto separate mixins;
* minor style/readability improvements;
* tests rewritten.

sibiryakov commented on frontera/contrib/backends/remote/messagebus.py:
```diff
@@ -41,41 +45,41 @@ def frontier_stop(self):
-                                      self._encoder.encode_offset(self.partition_id,
-                                                                  self.consumer.get_offset(self.partition_id)))
-        return requests
+                yield self._decoder.decode_request(encoded)
```
All users of this Backend have to be compatible with generators. Are they? Mainly Frontera Scrapy scheduler.

ZipFile replied:
Got the point, will check this. I believe, I've seen somewhere that get_next_requests should be iterable, not just list. From my point of view, making it iterable might open path to different kind of asynchronous backends in the future.

sibiryakov commented on frontera/contrib/backends/remote/messagebus.py:
```diff
@@ -93,19 +97,65 @@ def states(self):
+        return self._buffer.get_next_requests(max_n_requests, **kwargs)
+
+
+class AggregatePerHostMixin(object):
```
Is GenericMessageBusBackend usable without AggregatePerHostMixin? Please describe the use case.

ZipFile replied:
My use case is to use more complex aggregation than just by host. I need to aggregate both by client and by host in order to achieve my goals.

So yes, GenericMessageBusBackend won't be much use in distributed mode. It was supposed to provide basic blocks to build the whatever-you-want MB backends.

sibiryakov commented on frontera/contrib/backends/remote/messagebus.py:
```diff
@@ -93,19 +97,65 @@ def states(self):
+    per_host = defaultdict(list)
+
+    for request in requests:
+        key = get_host_fprint(request)
```
The above code skips the link, if fingerprint wasn't found in meta[b'domain'], but your implementation adds it with None key, which will cause MessageBus to bark.

ZipFile commented on frontera/contrib/backends/remote/messagebus.py:
```diff
@@ -41,41 +45,41 @@ def frontier_stop(self):
-        for host_fprint, host_links in six.iteritems(per_host):
-            self.spider_log_producer.send(host_fprint,
-                                          self._encoder.encode_add_seeds(host_links))
+        key = b'0123456789abcdef0123456789abcdef012345678'
```
There is no response nor request to calculate fprint from. The idea was not to perform aggregation at this point.

In overall, I dislike `b'0123456789abcdef0123456789abcdef012345678'` thing too, same goes for sending offsets after get_next_requests, since user might want to use different fprint calculation function. Most likely, I miss something important about how our message bus works, would be great if you could provide your opinion on what should be done here.

ZipFile commented:
Thanks for review, I'll try to update the code somewhere around weekend.

ZipFile commented:
New patchset arrived:

* Reverted change of the `get_next_requests` to generator. It appears, problem is much more complex than I expected. Fixing affected code is outside of this PR's scope;
* removed `GenericMessageBusBackend` with mixins in favor to configuration through manager settings;
* improved commit structure.

sibiryakov commented on tox.ini:
```diff
@@ -28,6 +28,7 @@ commands = flake8 setup.py frontera
 [flake8]
 ignore = E265,E501,F401,W391,W292,E226
 exclude = frontera/_version.py,versioneer.py,docs/source/conf.py,frontera/contrib/backends/opic/discovery.py
+application-import-names = frontera, tests
```
why is this change needed?

ZipFile replied:
Hints for `flake8-import-order`. Needed to suppress invalid import order errors.

sibiryakov commented on frontera/contrib/backends/remote/messagebus.py:
```diff
@@ -94,18 +117,20 @@ def states(self):
-    if b'fingerprint' not in request.meta[b'domain']:
+    try:
+        return request.meta[b'domain'][b'fingerprint']
+    except KeyError:
```
Side note:
I think we should hard-code UrlFingerprintMiddleware and DomainMiddleware during initialization in Manager. Without them Frontera isn't operational. So it doesn't make any sense to check for presence of this data, or check if these middlewares are enabled.

ZipFile replied:
Agreed, that was not clear for me too at first. Frontera in distributed mode totally unusable without those two.

sibiryakov commented on frontera/settings/default_settings.py:
```diff
@@ -26,6 +26,8 @@
 MAX_REQUESTS = 0
 MESSAGE_BUS = 'frontera.contrib.messagebus.zeromq.MessageBus'
 MESSAGE_BUS_CODEC = 'frontera.contrib.backends.remote.codecs.msgpack'
+MESSAGE_BUS_REQUEST_AGGREGATION_FUNCTION = 'frontera.contrib.backends.remote.messagebus.aggregate_per_host'
```
these methods needs to be documented in frontera-settings.rst

ZipFile replied:
Ok, will update the doc.

sibiryakov commented on tests/contrib/backends/remote/test_message_bus_backend.py:
```diff
@@ -0,0 +1,328 @@
+import tests
+
+
+patch = tests.mock.patch
```
looks like that on is unused?

ZipFile replied:
that's right

sibiryakov commented:
In general, it's OK. Only docs are needed.

sibiryakov commented:
Ok waiting for the docs @ZipFile

ZipFile commented:
Bit late, but docs are updated now.

sibiryakov commented on docs/source/topics/frontera-settings.rst:
```diff
@@ -238,6 +238,25 @@ Default: ``frontera.contrib.backends.remote.codecs.msgpack``
+
+Default: ``True``
+
+Enables use of :term:`overused buffer` in :class:`MessageBusBackend <frontera.contrib.backends.remote.messagebus.MessageBusBackend>`.
```
Is this term defined somewhere? If not please define it in glossary.rst

sibiryakov commented:
aggregation is outdated, this functionality must live in crawling strategy

sibiryakov commented:
optional overused buffer makes sense, but unlikely to be a problem in most cases. May be worth to live it as is, assuming that people who want to disable it are able to fork and customise Frontera.
